mmcv/runner/ipu/util.py

```diff
@@ -0,0 +1,324 @@
+        ipu_model_cfg (dict): a dictionary contains train_split_edges,
+            train_ckpt_nodes, see details in funcations model_sharding and
+            recomputation_checkpoint
+        fp16_cfg (dict): config for IPU fp16 training, currently support
+            configs: `loss_scale`, `velocity_accum_type`, `accum_type`.
+            see details in
+            https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/index.html
+    
+    Returns:
+        TrainEvalModel: IPU wrapped model.
+    """
+    ipu_model_cfg = ipu_model_cfg or {}
+    training = model.training if optimizer is not None else False
+    # set mixed-precision
+    if fp16_cfg is not None:
+        from mmcv.runner.fp16_utils import wrap_fp16_model
+        loss_scale = fp16_cfg['loss_scale']
+        wrap_fp16_model(model)
+        model.half()
+        # TODO tmp ussage to set loss scaling for torch original optimizer
+        if optimizer is not None:
+            optimizer.loss_scaling = loss_scale
+            if fp16_cfg.get('velocity_accum_type', False):
+                if fp16_cfg['velocity_accum_type'] == 'half':
+                    optimizer.velocity_accum_type = torch.half
+                else:
+                    optimizer.velocity_accum_type = torch.float32
+            if fp16_cfg.get('accum_type', False):
+                if fp16_cfg['accum_type'] == 'half':
+                    optimizer.accum_type = torch.half
+                else:
+                    optimizer.accum_type = torch.float32
+        # TODO support feature alignment for fp16
+        if len(modules_to_record) > 0:
+            raise NotImplementedError(
+                'Feature alignment for fp16 is not implemented')
+
+    # set model partition
+    if optimizer is None:
+        train_model = None
+    else:
+        # split model into multi-IPUs if specified
+        train_model = model_sharding(
+            copy.copy(model).train(),
+            ipu_model_cfg.get('train_split_edges', []))
+
+        recomputation_checkpoint(
+            train_model,
+            ipu_model_cfg.get('train_ckpt_nodes', []))
+
+        # TODO support feature alignment for gradient accumulation mode
+        if getattr(opts['training'].Training, 'gradient_accumulation', 1) > 1:
+            assert len(modules_to_record) == 0, \
+                'Feature alignment for grad-accumulation mode not implemented'
+
+        # TODO support feature alignment for multi-replica mode
+        if getattr(opts['training'], 'replication_factor', 1) > 1:
+            assert len(modules_to_record) == 0, \
+                'Feature alignment for multi-replica mode not implemented'
+
+    # split model into multi-IPUs if specified
+    assert len(ipu_model_cfg.get('eval_split_edges', [])) == 0,\
+        'Currently, BeginBlock can only be used once on the same model'
+    eval_model = model_sharding(copy.copy(model).eval(), ipu_model_cfg.get(
+        'eval_split_edges', []))
```

Comment:
If `eval_split_edges` should not be set, is it necessary to call `model_shareding`?